### PR TITLE
Add salt and pepper noise with docs and tests

### PR DESCRIPTION
**Pull request by vgilabert94**
I create a new PR because of some conflicts. (Last PR: https://github.com/kornia/kornia/pull/2733) 

#### Changes
Based on PR (https://github.com/kornia/kornia/pull/2612) and scikit-image implementation.
- Add random generator.
- Follow implementation of scikit image, but using element-wise operation (not indexing).

Fixes # (issue)
- same_on_batch: generate same parameter, but not same mask to apply. So is not same noise for all batch. Solved.

#### Type of change

- [ ] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Review by edgarriba — APPROVED** [review]
LGTM. Make sure that tests are passing well first in local. For docu you can use the `make doctest` command

**Comment by vgilabert94** [comment]
Docs test is resolved. 
About gradcheck test, I can't solve it, because of the cuda version I have installed (11.4) and the required by torch 1.2.1 (12.2). All gradcheck tests fail me locally. I can't update drivers, is not my computer where im working. 

`
UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11040). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org/ to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)
    return torch._C._cuda_getDeviceCount() >
`

Can someone solve it?

File: tests/augmentation/test_augmentation.py [hunk 1]
```diff
@@ -3209,6 +3210,99 @@ def test_smoke(self, device, dtype):
...
+        transform = RandomSaltAndPepperNoise(p=1.0)
+        output_tensor = transform(input_tensor)
+        assert input_tensor.shape[0] == output_tensor.shape[0]
+        assert input_tensor.shape[1] == output_tensor.shape[1]
+
+    def test_same_on_batch(self, device, dtype):
+        input_tensor = torch.ones(2, 1, 5, 5, device=device, dtype=dtype) * 0.5
+        transform = RandomSaltAndPepperNoise(p=1.0, same_on_batch=True)
+        output_tensor = transform(input_tensor)
+        assert_close(output_tensor[0], output_tensor[1])
```
Comment (johnnv1) [inline tests/augmentation/test_augmentation.py:3290]:
```suggestion
        self.assert_close(output_tensor[0], output_tensor[1])
```

File: tests/augmentation/test_augmentation.py [hunk 2]
```diff
@@ -3209,6 +3210,99 @@ def test_smoke(self, device, dtype):
...
+        torch.manual_seed(0)  # for random reproductibility
+        transform = RandomSaltAndPepperNoise(p=1.0)
+        input_tensor = torch.rand(1, 3, 16, 16, device=device, dtype=dtype)
+        input_tensor = utils.tensor_to_gradcheck_var(input_tensor)  # to var
+        assert gradcheck(
+            transform,
+            (input_tensor,),
+            raise_exception=True,
+            fast_mode=True,
+        )
```
Comment (johnnv1) [inline tests/augmentation/test_augmentation.py:3303]:
```suggestion
    def test_gradcheck(self, device):
        torch.manual_seed(0)  # for random reproductibility
        transform = RandomSaltAndPepperNoise(p=1.0)
        input_tensor = torch.rand(1, 3, 16, 16, device=device, dtype=torch.float64)
        self.gradcheck(transform, (input_tensor,))
```

**Comment by vgilabert94** [comment]
After last commits, I've created in google colab a times comparison of masking (multiply and add) and indexing algorithm. In both algorithm I use a random generator.

image

Looking at the results, Indexing seems to be fast than masking. I'd say it's worth upgrading to indexing in the next commit:
```
noisy_image = input.clone()
noisy_image[params["mask_salt"].to(input.device)] = 1.0
noisy_image[params["mask_pepper"].to(input.device)] = 0.0
```
What do you think?
